server/services/analyticsService.ts

```diff
@@ -335,3 +351,107 @@ export function compareCharacteristics(
 export function removeLevelPrefix(level: string): string {
   return /^[A-Z]+\.\s+(.*)\s*$/.exec(level)?.[1] || level
 }
+
+/**
+ * In-memory cache of the stiched tables values.
+ *
+ * The source tables are stored in S3, these files can be over 100MB
+ * and tranform them into a more useful format is also relatively "slow".
+ *
+ * It's wasteful to transfer these files from S3 and transform them from
+ * JSON to a matrix format for each request.
+ *
+ * A typical value would be ~30MB so small enough to be kept in memory.
+ * This has the advantage of clearing the cache if application is restarted.
+ *
+ * Values are cached based on source filename and list of columns.
+ * This means that potentially two graphs could use the same source table
+ * but different columns and the corresponding stiched tables would not clash.
+ *
+ * NOTE: The cached stiched tables are not filtered. The AnalyticsService
+ * would still filter them by prison or else. This filtering is relatively
+ * fast so we don't store all the small artefacts for each of the prisons.
+ */
+export class StitchedTablesCache {
+  /**
+   * A map-like object where the key is a unique string derived from
+   * the source table name and the columns (e.g. 'incentives_latest_narrow,prison,wing').
+   *
+   * The values contains the `expiresAt` date determining whether the value
+   * is expired or fresh.
+   *
+   * The `value` property contains the actual cached value, including
+   * the `stichedTable` matrix and the `date`/`modified` dates.
+   */
+  private static cache: Record<
+    string, // cacheKey [S3 object key, ...columnsToStitch].join()
```

Comment:
Should this comment be
`// cacheKey [TableType, ...columnsToStitch].join()`
?

Reply:
Comment updated ☺️